### ISSUE TITLE
[java] UseDiamondOperator false-positive (Lamda?) (7.0.0-rc4)

### ISSUE THREAD

**Issue by blacelle** [pmd/pmd#4821]
**Affects PMD Version:** 
7.0.0-rc4

**Rule:**

UseDiamondOperator

**Description:**

False-positive as the types are mandatory for proper type inference. Else, `enriched` is inferred to be `Map<String, ?>` and `enriched.putAll(properties);` is not allowed.

**Code Sample demonstrating the issue:**

```java
	private List<? extends Map<String, ?>> enrichFileConstrains(Map<String, ?> properties,
			List<Map<String, ?>> documentFields) {
		return documentFields.stream().map(m -> new LinkedHashMap<String, Object>(m)).map(enriched -> {

			enriched.putAll(properties);

			return enriched;
		}).collect(Collectors.toList());
	}
```

**Expected outcome:**

PMD reports a violation at line ..., but that's wrong. That's a false positive.

**Running PMD through:** Maven

**Comment by adangel** [comment]
If you help the compiler and tell, which of the `?` are actually the same type, then you can use the diamond operator:

```java
    private <T> List<? extends Map<String, T>> enrichFileConstrains2(Map<String, T> properties, 
                                                  List<Map<String, T>> documentFields) {
        return documentFields
                    .stream()
                    .map(m -> new LinkedHashMap<>(m))
                    .map(enriched -> { 
                        enriched.putAll(properties);
                        return enriched;
                    })
                    .collect(Collectors.toList());
    }
```
